pgrx/src/datum/array.rs

```diff
@@ -503,6 +653,52 @@ fn as_slice<'a, T: Sized>(array: &'a Array<'_, T>) -> Result<&'a [T], ArraySlice
     Ok(slice)
 }
 
+#[inline(always)]
+fn as_mut_slice<'a, T: Sized>(array: &'a mut Array<'_, T>) -> Result<&'a mut [T], ArraySliceError> {
+    if array.contains_nulls() {
+        return Err(ArraySliceError::ContainsNulls);
+    }
+
+    let slice =
+        unsafe { std::slice::from_raw_parts_mut(array.raw.data_ptr() as *mut _, array.len()) };
+    Ok(slice)
+}
+
+/// Creates an `Array<'a, T>` with zero-elements
+/// Slightly faster than new_array_with_len(0)
+pub fn new_empty_array<'a, T: Sized>() -> Result<Array<'a, T>, ArrayAllocError>
+where
+    T: IntoDatum,
+    T: UnboxDatum,
+{
+    unsafe {
+        let raw_array = RawArray::new_empty_array_type::<T>()?;
+        let datum: pgrx_pg_sys::Datum = pg_sys::Datum::from(raw_array.into_ptr().as_ptr());
+        Array::<'a, T>::from_polymorphic_datum(datum, false, pg_sys::get_array_type(T::type_oid()))
+            .ok_or(ArrayAllocError::MemoryAllocationFailed)
+    }
+}
+
+/// Creates an `Array<T>` of a fixed len, with 0 for all elements
+/// Uses a single PG allocation rather than
+#[inline(always)]
+pub fn new_array_with_len<'a, T: Sized>(len: usize) -> Result<Array<'a, T>, ArrayAllocError>
+where
+    T: IntoDatum,
+    T: UnboxDatum,
```

Comment:
I do not believe `T: IntoDatum + UnboxDatum + Sized` is a sufficient bound to establish that it is sound to create that type with all-zero initialization. This function is probably `unsafe` in actuality, albeit with a maybe-trivial precondition.

Reply:
replaced with an opt in ArrayFastAllocSubType trait, types are specifically opted in